### PR TITLE
[WIP] Trainer wrapper for image classification model

### PR DESCRIPTION
**Pull request by rahul-tuli**
The Goal of this PR is to refactor logic and make training image classification models easier, 
Right now this PR is a work in progress, and any ideas on a more extensible design are welcome

```python
trainer = ImageClassificationTrainer(
            checkpoint_path = 'models/resnet50.pth',
            recipe_path='recipes/my-recipe.yaml',
            dataset = 'imagenette',
            dataset_path = 'dataset',
            train_batch_size= 32,
            test_batch_size = 64,

)
trainer.run_training()
```

### CODE REVIEW

File: src/sparseml/pytorch/models/registry.py [hunk 1]
```diff
@@ -207,8 +207,11 @@ def register(
         if not isinstance(key, List):
             key = [key]
 
-        def decorator(const_func):
-            wrapped_constructor = ModelRegistry._registered_wrapper(key[0], const_func)
+        def decorator(constructor_func):
```
Comment (bfineran) [inline src/sparseml/pytorch/models/registry.py:210]: confirming that the changes here are just for readability?

File: src/sparseml/pytorch/utils/trainer.py [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ImageClassificationTrainer:
+    "Utility class for training image classification models"
+
+    def __init__(
+            self,
+            checkpoint_path: str,
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/trainer.py:48]: we definitely need to cut down on the number of arguments we have here. for example, the dataloaders should be able to be constructed outside of the trainer and then passed in. Additionally, let's also keep model saving outside of the trainer for now. and can you remind me on the blocks we saw for model construction needing to be in the trainer class?

File: src/sparseml/pytorch/utils/trainer.py [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+        self.device = device
+        self.debug_steps = debug_steps
+        self.save_best_after = save_best_after
+        self.save_epochs = save_epochs
+
+        self.dataset_kwargs = dataset_kwargs or {}
+        self.optim_kwargs = optim_kwargs or {}
+
+        self.world_size = int(os.environ.get("WORLD_SIZE", 1))
+        self.env_rank = int(os.environ.get("RANK", -1))
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/trainer.py:103]: DDP should be handled outside of the trainer, if possible, is there any reason it has to be in here? most of the rank specific calls should happen outside the trainer, right?

**Review by bfineran — CHANGES_REQUESTED** [review]
left some initial comments, net net, can we move stuff like dataloader construction, model saving, and DDP logic to outside of the Trainer? As it is, there are too many arguments specific to our IC train script

File: src/sparseml/pytorch/models/registry.py (hunk 1 above)
Comment (rahul-tuli) [reply src/sparseml/pytorch/models/registry.py:210]: yes!

**Pull request closed without merging**